frontend/stepper/index.ts

```diff
@@ -728,74 +755,103 @@ function* stepperEnabledSaga(app: App) {
     currentStepperTask = yield* fork(app.stepperApi.rootStepperSaga, app);
 }
 
-export function* stepperDisabledSaga(action, leaveContext = false) {
+export function* stepperDisabledSaga(action, leaveContext = false, clearSourceHighlight = true) {
     /* Cancel the stepper task if still running. */
     const oldTask = currentStepperTask;
-    console.log('try to disable stepper', oldTask, leaveContext);
+    log.getLogger('stepper').debug('try to disable stepper', oldTask, action, leaveContext, clearSourceHighlight, arguments);
 
     if (leaveContext) {
         yield* put(taskResetDone(false));
     }
 
+    if (clearSourceHighlight) {
+        yield* call(clearSourceHighlightSaga);
+    }
+
     if (oldTask) {
+        yield* put({type: ActionTypes.StepperTaskCancelled});
+
         // @ts-ignore
         yield* cancel(oldTask);
 
-        yield* put({type: ActionTypes.StepperTaskCancelled});
+        // Warning: put no code after this, because the cancel will instantly kill this saga
     }
-
-    yield* call(clearSourceHighlightSaga);
 }
 
-function* stepperInteractBeforeSaga(app: App, {meta: {resolve, reject}}) {
+function* stepperInteractBeforeSaga(app: App, {payload: {stepperContext}, meta: {resolve, reject}}: {payload: {stepperContext: StepperContext}, meta: {resolve: any, reject: any}}) {
     let state: AppStore = yield* select();
 
-    console.log('inside stepper interact before');
-
     /* Has the stepper been interrupted? */
     if (isStepperInterrupting(state) || StepperStatus.Clear === state.stepper.status) {
-        console.log('stepper is still interrupting');
+        log.getLogger('stepper').debug('stepper is still interrupting');
         yield* call(reject, new StepperError('interrupt', 'interrupted'));
 
         return;
     }
 
+    // Update speed if we use speed
+    const context = quickAlgoLibraries.getContext(null, state.environment);
+    let newDelay = 0;
+    if ('main' === state.environment) {
+        if (null !== stepperContext.speed && undefined !== stepperContext.speed) {
+            stepperContext.speed = getStepper(state).speed;
+            newDelay = stepperMaxSpeed - stepperContext.speed;
+        }
+        // log.getLogger('stepper').debug('stepper interact before background run data', stepperContext.backgroundRunData);
+        if (stepperContext.backgroundRunData && stepperContext.backgroundRunData.steps) {
+            const runData = stepperContext.backgroundRunData;
+            // if (runData.result || (!runData.result && runData.steps && runData.steps >= Codecast.runner._steps + 10)) {
+            //     newDelay = newDelay / 4;
+            // }
+            const t = Codecast.runner._steps / runData.steps;
+            const y0 = newDelay;
+            const y1 = newDelay / 40;
+            const y2 = newDelay / 40;
+            const y3 = newDelay;
+
+            newDelay = (1-t)*((1-t)*((1-t)*y0+t*y1)+t*((1-t)*y1+t*y2))+t*((1-t)*((1-t)*y1+t*y2)+t*((1-t)*y2+t*y3));
```

Comment:
Maybe a comment explaining the logic between the formula?

I understand it's the formula which accelerates and decelerates over time from newDelay to newDelay / 40, but a comment could help.

Reply:
Ok, I've added a comment to explain :+1: